basic support for multiple conditions added

preda2or opened the pull request:
Fixes issue #778 

created basic support for multiple conditions using logic anyOneTrue or allTrue from a list of conditions. 

- Need suggestions for UI improvements. 
- One problem i have noticed is that indenting breaks the data variable field of AnyVariableAndDataPair in the custom inspector.
- For backward compatibility setting the anyVar field to null doesn't seem to work. it seems to get reinitialized on changing the number of condition. so for now I am checking if anyVar.variable != null to skip backward compatibility.

stevehalliwell commented:
Had a quick look over it, seems like a good start, some coding standard stuff to fix up but that's not a big deal at the moment.

Some quick suggestions to chase up.
- Use OnValidate to force list to be at least 1 long, empty condition list makes no sense
- Gaps between conditions, via Space or Separator
- RE; AnyVar not nulling, try putting [HideInInspector] on it, the serialized object may still be defaulting it even though the custom editor isn't showing it.
- Definitely do indent, the data drawerer still functions but does have a visual error, I'll see I can fix that.
- Add a demo scene somewhere in FungusExamples to demonstrate and validate functionality.
- The summary will end up being the pseudo code for the condition chain, but it may not fit easily in the inspector space, I would try putting that in a label above or below the rest of the command for users to cross check against.

If you are still wanting to extend it to have A && B || C, then having that stored in your ConditionExpression class makes sense, just don't draw that element on the last item in the list. If you are going to include that feature, you'll definitely want to look at using reorderable list for the ConditionExpression list though

preda2or commented:
Addressing your issues in order

1. In OnValidate if the list is empty i have added an empty condition to it.
2. Added some space between conditions.
3. I tried adding [HideInInspector] to it, but it is still not nulling. Is there a problem with checking anyVar.variable == null as i am doing right now?
4. Added the code for indenting but it's commented out for now because if i indent the data drawer has a visual glitch as well as the drop down isn't working. You can un-comment that when you fix it.
5. Added a demo scene(MultipleConditionsExample).
6. I'm not sure which summary you are referring to here. I have already changed the GetSummary() of the VariableCondition class in my previous commit. It displays next to the if or else if command in the command list window itself.

I am not planning to support conditions like A && B || C directly in the if or else if commands because that requires brackets/priority. It can be done in a separate command though. But as you said people who want that kind of behavior will probably use lua if.

stevehalliwell commented on Assets/Fungus/Scripts/Commands/VariableCondition.cs:
```diff
@@ -6,49 +6,147 @@
-    public abstract class VariableCondition : Condition, ISerializationCallbackReceiver
+
+    [System.Serializable]
+    public class conditionExpression
```
Should start with Capital and could use  a class summary

stevehalliwell commented on Assets/Fungus/Scripts/Commands/VariableCondition.cs:
```diff
@@ -6,49 +6,147 @@
+    public abstract class VariableCondition : Condition, ISerializationCallbackReceiver
+    {       
+
+        [Tooltip("The type of comparison to be performed")]
```
Prefer explaining what the two modes mean/ what they will do.

stevehalliwell commented on Assets/Fungus/Scripts/Commands/VariableCondition.cs:
```diff
@@ -6,49 +6,147 @@
+            }
+            else
+            {
+                connector = " And ";
```
These don't read easily in the command summary preview, try all caps and maybe wrap in bold tags

stevehalliwell commented on Assets/Fungus/Scripts/Editor/VariableConditionEditor.cs:
```diff
@@ -33,61 +33,90 @@ public class VariableConditionEditor : CommandEditor
+
+
+                //uncomment this for indentation
+                // EditorGUI.indentLevel++;
```
Pushed a fix for the data drawer to develop

stevehalliwell commented on Assets/Fungus/Scripts/Editor/VariableConditionEditor.cs:
```diff
@@ -33,61 +33,90 @@ public class VariableConditionEditor : CommandEditor
-            if (selectedVariable != null)
-            {
-                compareOperatorProp.enumValueIndex = selectedIndex;
+                GUILayout.Label("Condition "+(i+1).ToString(),EditorStyles.boldLabel);
```
Some kind of heading before the condition list would be a good, but a heading per condition seems overkill. Ends up taking a lot of space.

stevehalliwell commented on Assets/Fungus/Scripts/Commands/VariableCondition.cs:
```diff
@@ -6,49 +6,147 @@
-            string summary = anyVar.variable.Key + " ";
-            summary += VariableUtil.GetCompareOperatorDescription(compareOperator) + " ";
-            summary += anyVar.GetDataDescription();
+            string summary = "";
```
Given how many += end up being done here and given following loop. It's probably worth moving this to a StringBuilder.

preda2or commented:
Made your requested changes

stevehalliwell commented:
Checked out locally to test out, found a problem in back compat. When updating a command from old fields to new version, it's ending up adding it multiple of the same condition. Due to not nulling out the existing old data after moving to new.

stevehalliwell commented:
I'm going to pull into a local branch and make the required adjustments before merging into develop.
